Cruise Control deployment (#2773)
strimzi/strimzi-kafka-operator: documentation/modules/cruise-control/proc-deploying-cruise-control.adoc

// This module is included in the following assemblies:
//
// assembly-cruise-control-concepts.adoc

[id='proc-deploying-cruise-control-{context}']
= Deploying Cruise Control

Deploy Cruise Control to your {ProductName} cluster through the `cruiseControl` property in the `Kafka` resource.

> Refer to the xref:ref-sample-kafka-resource-config-deployment-configuration-kafka[sample Kafka YAML configuration] for an overview of the `Kafka` resource and its properties.

Reviewer: ```suggestion
For an overview of the `Kafka` resource, see the xref:ref-sample-kafka-resource-config-deployment-configuration-kafka[sample Kafka YAML configuration].
```

> You can configure `cruiseControl` properties as part of a deployment or redeployment of a Kafka cluster.

Reviewer: ```suggestion
You can configure `cruiseControl` properties as part of a deployment or redeployment of a Kafka cluster. Deploy one instance of Cruise Control per Kafka cluster.
```

+
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: Kafka
metadata:
  name: my-cluster
spec:
  # ...
  cruiseControl:
    replicas: 1 <1>
    capacity: <2>
      networkIn: 10000
      networkOut: 10000
      # ...
    config: <3>
      default.goals: >
         com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,
         com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal
      cpu.balance.threshold: 1.1
      metadata.max.age.ms: 300000
      send.buffer.bytes: 131072
      # ...
    resources: <4>
      requests:
        cpu: 200m
        memory: 64Mi
      limits:
        cpu: 500m
        memory: 128Mi
    logging: debug <5>
    template: <6>
      pod:
        metadata:
          labels:
            label1: value1
        securityContext:
          runAsUser: 1000001
          fsGroup: 0
        terminationGracePeriodSeconds: 120
    readinessProbe: <7>
      initialDelaySeconds: 15
      timeoutSeconds: 5
    livenessProbe: <8>
      initialDelaySeconds: 15
      timeoutSeconds: 5
# ...
----
<1> Number of Cruise Control application instances; set to `1` to run or `0` to pause the application.

> <2> Specifies the Cruise Control capacity configuration. For more information, see xref:capacity_configuration[capacity configuration].

Reviewer: ```suggestion
<2> Specifies capacity limits for broker resources. For more information, see xref:capacity_configuration[Capacity configuration].
```

> <3> Specifies the Cruise Control configuration. You can provide any xref:ref-cruise-control-configuration-{context}[standard configuration option] apart from those managed directly by {ProductName}.

Reviewer: ```suggestion
<3> Defines the Cruise Control configuration, including default and enabled optimization goals. You can provide any xref:ref-cruise-control-configuration-{context}[standard configuration option] apart from those managed directly by {ProductName}.
```

<4> CPU and memory resources reserved for Cruise Control. For more information, see xref:assembly-resource-limits-and-requests-deployment-configuration-kafka[].

> <5> Logging configuration, to log messages with a given severity (debug, info, warn, error, fatal) or above.

Reviewer: ```suggestion
<5>Defines the logging configuration. Messages are logged with a severity level of `debug`, `info`, `warn`, `error`, or `fatal`.
```

Author: I incorrectly  defined the logging configuration here, I have just updated the docs with the correct configuration. Take a look and let me know what you think!

Reviewer: New version looks good.

<6> xref:assembly-customizing-deployments-str[Customization of deployment templates and pods].
<7> xref:assembly-healthchecks-deployment-configuration-kafka[Healthcheck readiness probes].

> <8> xref:assembly-healthchecks-deployment-configuration-kafka[Healthcheck liveness probes].

Reviewer 2: Does this belong in a procedure, rather than a concept?

Author: I think it could be separated into a reference but have been shuffling these around. Let me get the final word from @laidan6000

Reviewer: On balance, I think it should stay here because the example configuration is helpful within the context of the task. In other words, it doesn't seem to stand alone as a concept or reference module.

. Create or update the resource:
+
[source,shell,subs="+quotes"]
----
kubectl apply -f _kafka.yaml_
----

> . Verify the Cruise Control deployment:

Reviewer: ```suggestion
. Verify that Cruise Control was successfully deployed:
```

For an overview of the `Kafka` resource, see the xref:ref-sample-kafka-resource-config-deployment-configuration-kafka[sample Kafka YAML configuration].

You can configure `cruiseControl` properties as part of a deployment or redeployment of a Kafka cluster.
Deploy one instance of Cruise Control per Kafka cluster.

.Prerequisites

* A Kubernetes cluster
* A running Cluster Operator

.Procedure

. Edit the `cruiseControl` property of the `Kafka` resource.
+
The properties you can configure are shown in this example configuration:
+
[source,yaml,subs="attributes+"]
----
apiVersion: {KafkaApiVersion}
kind: Kafka
metadata:
  name: my-cluster
spec:
  # ...
  cruiseControl:
    replicas: 1 <1>
    capacity: <2>
      networkIn: 10000KB/s
      networkOut: 10000KB/s
      # ...
    config: <3>
      default.goals: >
         com.linkedin.kafka.cruisecontrol.analyzer.goals.RackAwareGoal,
         com.linkedin.kafka.cruisecontrol.analyzer.goals.ReplicaCapacityGoal
      cpu.balance.threshold: 1.1
      metadata.max.age.ms: 300000
      send.buffer.bytes: 131072

>       # ...

Reviewer: @kyguy - Does it make sense to include a `goals` configuration here too?

<4> CPU and memory resources reserved for Cruise Control. For more information, see xref:assembly-resource-limits-and-requests-deployment-configuration-kafka[].
<5> Defined loggers and log levels added directly (inline) or indirectly (external) through a ConfigMap. A custom ConfigMap must be placed under the log4j.properties key. Cruise Control has a single logger called cruisecontrol.root.logger. You can set the log level to INFO, ERROR, WARN, TRACE, DEBUG, FATAL or OFF. For more information, see xref:logging_configuration[Logging configuration].
<6> xref:assembly-customizing-deployments-str[Customization of deployment templates and pods].
<7> xref:assembly-healthchecks-deployment-configuration-kafka[Healthcheck readiness probes].
<8> xref:assembly-healthchecks-deployment-configuration-kafka[Healthcheck liveness probes].

. Create or update the resource:
+
[source,shell,subs="+quotes"]
----
kubectl apply -f _kafka.yaml_
----

. Verify that Cruise Control was successfully deployed:
+
[source,shell,subs="+quotes"]
----
kubectl get deployments -l app.kubernetes.io/name=strimzi
----

.What to do next

> After configuring and deploying Cruise Control, you can xref:proc-interacting-with-cruise-control-api-{context}[interact with the Cruise Control API]

Reviewer: @kyguy - Please comment this out because the target procedure is not ready. Also, I think the best next step is to generate optimization proposals and use them to rebalance the cluster.
